src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -0,0 +1,1211 @@
+//include header file
+#include "hiopAlgPrimalDecomp.hpp"
+#include "hiopInterfacePrimalDecomp.hpp"
+
+#include <cstring>
+#include <cassert>
+
+namespace hiop
+{
+#ifdef HIOP_USE_MPI
+  /* This struct provides the info necessary for the recourse approximation function
+   * buffer[n+1] contains both the function value and gradient w.r.t x
+   * buffer[0] is the function value and buffer[1:n] the gradient
+   * Contains send and receive functionalities for the values in buffer
+   */
+  struct ReqRecourseApprox
+  {
+    ReqRecourseApprox() : ReqRecourseApprox(1) {}
+    ReqRecourseApprox(const int& n)
+    {
+      n_=n;
+      buffer=new double[n_+1];
+    }
+
+    int test() 
+    {
+      int mpi_test_flag; MPI_Status mpi_status;
+      int ierr = MPI_Test(&request_, &mpi_test_flag, &mpi_status);
+      assert(MPI_SUCCESS == ierr);
+      return mpi_test_flag;
+    }
+    void post_recv(int tag, int rank_from, MPI_Comm comm)
+    {
+      int ierr = MPI_Irecv(buffer, n_+1, MPI_DOUBLE, rank_from, tag, comm, &request_);
```

Comment:
Method `buffer` in `MpiBufferManager` class returns pointer to buffer data.
```suggestion
      int ierr = MPI_Irecv(bm_->buffer, n_+1, MPI_DOUBLE, rank_from, tag, comm, &request_);
```